runtime/eventlog.c

```diff
@@ -0,0 +1,257 @@
+#define CAML_INTERNALS
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <inttypes.h>
+#include "caml/alloc.h"
+#include "caml/eventlog.h"
+#include "caml/osdeps.h"
+
+#define CTF_MAGIC 0xc1fc1fc1
+#define CAML_TRACE_VERSION 0x1
+
+struct ctf_stream_header {
+  uint32_t magic;
+  uint16_t caml_trace_version;
+  uint16_t stream_id;
+};
+
+static struct ctf_stream_header header = {
+  CTF_MAGIC,
+  CAML_TRACE_VERSION,
+  0
+};
+
+#pragma pack(1)
+struct ctf_event_header {
+  uint64_t timestamp;
+  uint32_t pid;
+  uint32_t id;
+};
+
+static uintnat alloc_buckets [20] =
+  {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+
+
+struct event {
+  struct ctf_event_header header;
+  uint8_t  phase; /* for GC events */
+  uint8_t  counter_kind; /* misc counter name */
+  uint8_t  alloc_bucket; /* for alloc counters */
+  uint32_t count; /* for misc counters */
+};
+
+static FILE* output;
+static uintnat eventlog_startup_timestamp = 0;
+static uint32_t eventlog_startup_pid = 0;
+static uintnat eventlog_paused = 0;
+
+uintnat caml_eventlog_enabled = 0;
+
+#define EVENT_BUF_SIZE 4096
+struct event_buffer {
+  uintnat ev_generated;
+  struct event events[EVENT_BUF_SIZE];
+};
+
+static struct event_buffer* evbuf;
+
+void setup_evbuf()
+{
+  CAMLassert(!evbuf);
+  evbuf = malloc(sizeof(*evbuf));
+
+  if (!evbuf) return;
+
+  evbuf->ev_generated = 0;
+}
+
+void setup_eventlog_file()
+{
+  char *filename;
+  char *ocaml_eventlog_filename;
+
+  ocaml_eventlog_filename = caml_secure_getenv("OCAML_EVENTLOG_FILE");
+  if (ocaml_eventlog_filename) {
+    filename = malloc(strlen(ocaml_eventlog_filename) + 64);
+    sprintf(filename, "%s.%d.eventlog",
+            ocaml_eventlog_filename, eventlog_startup_pid);
```

Comment:
The size handling in this code is approximative, and we got rid of all calls to unsafe C functions such as `sprintf` in 774e30e138dc22a5acd6cfac03ae25194ae8cd6e. It would be better to avoid `sprintf` here, and I suppose that using `caml_alloc_sprintf` would be fine.

Reply:
Thanks! Sorry, I'm unsure how one would use `caml_alloc_sprintf` in the runtime internally. Should the resulting value be passed to fopen by using `caml_stat_strdup` and then freed with `caml_stat_free`?
Thank you very much. :)

Reply:
I'm not an expert in string programming in the runtime either, but:
- the result of `caml_alloc_sprintf` is an OCaml string on the OCaml heap; you don't need to deallocate it (it's gc-ed) but you would need to use `String_val(foo)` to get a C string out of it
- I suspect that you need to use `fopen_os` as defined in caml/misc.h instead of `fopen`, for reasons of Windows unicode stuff stuff. The argument is a `char_os *`, and you would get it by calling `caml_stat_strdup_to_os(String_val(sprintf_result))`; then free it with `caml_stat_free` instead.

For the `_os` stuff, see the [Manual > C interface > Windows Unicode APIs](http://caml.inria.fr/pub/docs/manual-ocaml/intfc.html#sec483).

Reply:
I'm a little late to this conversation, but: you don't need to allocate OCaml strings on the OCaml heap here. It's probably simpler to just change uses of `sprintf` to `snprintf` (this is what 774e30e does). `snprintf` has a very similar API to `sprintf`, but takes an explicit length parameter.